### ISSUE TITLE
Rule proposal: wrong array index after length check (?)

### ISSUE THREAD

**Issue by jendrikw** [typescript-eslint/typescript-eslint#12242]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

This rule would apply to values of type `Array<T>`. After checking the length, an invalid array index should be flagged as "is always undefined".

I think this is not solvable by better typechecking from typescripts's side, because user could possibly create strange subclasses of Array, which I expect to be rare in practice.

### Fail Cases

```TypeScript
arr.length === 0 && arr[0] === "foo"
arr.length <= 5 && arr[5] === "foo";
```

### Pass Cases

```TypeScript
arr.length === 1 && arr[0] === "foo";
arr.length >= 5 && arr[4] === "foo";
```

### Additional Info

I had a bug where I checked against 0 instead of 1.

**Comment by bradzacher** [comment]
> My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.

This proposal fails this check - it does not require type information, nor is it checking TS syntax.

You mention "This rule would apply to values of type Array<T>." but there's no reason the rule would specifically require this type constraint given its checking the combination of both a `.length` property AND index access - the two together are unique enough to not require types.

> I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal). 

This proposal also fails this case - it's a pretty niche bug vector overall.

Finally there's various complexities to this - properly handling the control-flow analysis and scope analysis for this is fairly complex and difficult to build and maintain. So without it being a high-value thing to check for the burden isn't justified.

For these reasons I am going to reject this proposal.

**Comment by jendrikw** [comment]
I understand. Thank you for taking your time.
